oak/server/grpc_event.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef OAK_GRPC_SERVER_GRPC_EVENT_H_
+#define OAK_GRPC_SERVER_GRPC_EVENT_H_
+
+#include "grpc_stream.h"
+
+namespace oak {
+namespace grpc_server {
+
+// Representation of a gRPC event.
+//
+// An event object is added to the gRPC completion queue once an event is
+// completed. Then handle() is called to invoke any subsequent dependent
+// work.
+class BaseGrpcEvent {
+ public:
+  virtual void handle() = 0;
+
+ protected:
+  BaseGrpcEvent(std::shared_ptr<GrpcStream> stream_) : stream_(stream_) {}
+  std::shared_ptr<GrpcStream> stream_;
```

Comment:
Alternatively we may just pass the stream as  an argument to `handle()` and avoid keeping it around here and complicating the constructor? Effectively this class is just representing a lambda / callback function, so it would be good if that's obvious from looking at it, even if the two alternatives are otherwise practically equivalent. In fact perhaps we can even avoid the base class entirely, and just declare a new type for the callback as `void callback(GrpcStream *stream)` or something like that, or perhaps `std::function<void(GrpcStream*)>` (not entirely sure of the C++11 syntax for functions, see https://en.cppreference.com/w/cpp/utility/functional/function). That would also allow us to alternatively pass a function reference or even a lambda, which may or may not make the code more readable.

Reply:
The reason why the stream is on the event object rather than passed in is because there can be multiple concurrent streams and the only thing keeping track of the stream state is the event objects.

std::function is an object rather than a function pointer so still requires additional allocation in order to get a pointer we can use as a tag. It can be extended as a class but I'm not sure it improves readability because we already have the stream class member so can't use simple lambda form and if you extend it then you need to implement operator() which isn't necessarily more readable than a normal class with handle()